domain/modelagent/state/modelstate.go

```diff
@@ -164,6 +166,52 @@ func (st *State) checkUnitNotDead(
 	}
 }
 
+// GetMachineCountNotUsingBase returns the number of machines that are not
+// using one of the supplied bases. If no machines exist in the model or if
+// no machines exist using a different base, zero is returned with no error. If
+// a empty set of bases is provided every machine in the model will be included
+// in the count.
+func (st *State) GetMachineCountNotUsingBase(
+	ctx context.Context,
+	bases []corebase.Base,
+) (int, error) {
+	db, err := st.DB()
+	if err != nil {
+		return 0, errors.Capture(err)
+	}
+
+	machineBaseValues := make(machineBaseValues, 0, len(bases))
+	for _, base := range bases {
+		machineBaseValues = append(machineBaseValues, base.String())
+	}
+	machineCount := machineCount{}
+
+	stmt, err := st.Prepare(`
+SELECT count(*) AS &machineCount.count
+FROM   machine
+WHERE  base NOT IN ($machineBaseValues[:])
+`,
+		machineBaseValues, machineCount)
+	if err != nil {
+		return 0, errors.Capture(err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err := tx.Query(ctx, stmt, machineBaseValues).Get(&machineCount)
+		if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+			return err
```

Comment:
Capture or errors.Errorf ?

Reply:
This offers no value. It is within a closure where the caller does this task. Doing this would only waste time in the transaction for no more context. Most txn code we have does this where it can be reasonably inferred like this case.